In [15]:
import pandas as pd
import numpy as np
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.ensemble import AdaBoostClassifier

In [16]:
df = pd.read_csv('model_data.csv')

In [17]:
df.drop(columns='Unnamed: 0',axis=1,inplace=True)

In [18]:
df.head()

Gender  MaritalStatus  Travelling  Role  OverTime  Qualifications  \
0       0              2           2     7         1               3   
1       1              1           1     7         0               4   
2       0              2           2     5         1               3   
3       1              1           2     0         0               5   
4       0              1           2     5         0               3   

   Satisfaction  Engagement  Level  Work Experience  TrainingTimesLastYear  \
0             4           2      2                1                      1   
1             4           3      2                9                      2   
2             4           3      4                5                      2   
3             4           3      3                1                      3   
4             1           1      4                2                      3   

   YearsInCurrentRole  YearsSinceLastPromotion       Age     Daily   Monthly  \
0                   1                        1  0.308288 -1.487720 -0.090531   
1                   7                        7  0.938585 -1.475988 -0.367020   
2                   5                        0  1.253734  0.401122  2.545160   
3                   1                        1 -0.847256  1.027608  0.214365   
4                   2                        0  1.778981  1.677557  2.330740   

   MonthlyRate  SalaryHike  DistanceFromHome  Turnover  
0    -1.332782    2.064153         -0.892559         0  
1     1.330416    0.754004          1.682756         0  
2     1.557335   -0.294115          0.456415         0  
3    -1.519952    1.016034          0.456415         0  
4    -0.179357   -0.556145         -1.015193         0

In [19]:
X=df.drop(columns='Turnover')
Y = df['Turnover']

#### Model Selection

In [24]:
## Random forest
model = RandomForestClassifier()

n_estimators = [100,200,500,1000]
# Minimum number of samples required to split a node
min_samples_split = [2, 5, 10]
# Minimum number of samples required at each leaf node
min_samples_leaf = [1, 2, 4]
# Method of selecting samples for training each tree
bootstrap = [True, False]
# Create the random grid
grid = {'n_estimators': n_estimators,
               'min_samples_split': min_samples_split,
               'min_samples_leaf': min_samples_leaf,
               'bootstrap': bootstrap}



In [25]:
rand_cv = GridSearchCV(model,grid,cv=5 )

In [26]:
rand_cv.fit(X,Y)

GridSearchCV(cv=5, estimator=RandomForestClassifier(),
             param_grid={'bootstrap': [True, False],
                         'min_samples_leaf': [1, 2, 4],
                         'min_samples_split': [2, 5, 10],
                         'n_estimators': [100, 200, 500, 1000]})

In [27]:
best_parameters = rand_cv.best_params_
print(best_parameters)
# higest accuracy

highest_accuracy = rand_cv.best_score_
print(highest_accuracy)

{'bootstrap': True, 'min_samples_leaf': 4, 'min_samples_split': 10, 'n_estimators': 500}
0.8238095238095239


In [29]:
model = GradientBoostingClassifier()

learning_rate=[0.05,0.10,0.20,0.25,0.3]
n_estimators=[100,400,600,800,1000]
max_depth=[1,2,3,4,5]


grid1 = {'learning_rate': learning_rate,
               'n_estimators': n_estimators,
               'max_depth': max_depth}



In [30]:
GBoost = GridSearchCV(model,grid1,cv=5)

In [31]:
GBoost.fit(X,Y)

GridSearchCV(cv=5, estimator=GradientBoostingClassifier(),
             param_grid={'learning_rate': [0.05, 0.1, 0.2, 0.25, 0.3],
                         'max_depth': [1, 2, 3, 4, 5],
                         'n_estimators': [100, 400, 600, 800, 1000]})

In [32]:
best_parameters = GBoost.best_params_
print(best_parameters)
# higest accuracy

highest_accuracy = GBoost.best_score_
print(highest_accuracy)

{'learning_rate': 0.1, 'max_depth': 1, 'n_estimators': 800}
0.825170068027211


In [33]:
model = AdaBoostClassifier()



n_estimators=[30,50,100,200]
learning_rate=[0.5,1.0,1.5,2.0] 
algorithm=['SAMME', 'SAMME.R']

grid3 = {'n_estimators': n_estimators,
               'learning_rate': learning_rate,
               'algorithm': algorithm}


In [34]:
ADBoost = GridSearchCV(model,grid3,cv=5)

In [35]:
ADBoost.fit(X,Y)

GridSearchCV(cv=5, estimator=AdaBoostClassifier(),
             param_grid={'algorithm': ['SAMME', 'SAMME.R'],
                         'learning_rate': [0.5, 1.0, 1.5, 2.0],
                         'n_estimators': [30, 50, 100, 200]})

In [36]:
best_parameters = ADBoost.best_params_
print(best_parameters)
# higest accuracy

highest_accuracy = ADBoost.best_score_
print(highest_accuracy)

{'algorithm': 'SAMME', 'learning_rate': 0.5, 'n_estimators': 50}
0.8272108843537416


#### best performing model from the three ensemble models is ADABOOST wiht accuracy of 82.7%
*with {'algorithm': 'SAMME', 'learning_rate': 0.5, 'n_estimators': 50}* 